### PR TITLE
CLN: Replace FrameOrSeriesUnion annotation by DataFrame | Series

### LINKED ISSUE #40876: STYLE, TYP frame-or-series-union check no longer relevant
This check is no longer relevant:

https://github.com/pandas-dev/pandas/blob/9ab55b4aed9018b070c939ffbdf232f99bf50f1a/.pre-commit-config.yaml#L83-L88

Since using PEP604 rewrites, this would be written as
```
x: Series | DataFrame
```
anyway.
```
x: FrameOrSeriesUnion
```
is actually just as long as the above, and arguably less explicit.

----

@simonjayhawkins thoughts on replacing all the `FrameOrSeriesUnion` occurrences, removing the alias, and removing the above check?

### PR DESCRIPTION
**Pull request by Swanand01**
- [x] closes #40876
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

### CODE REVIEW

**Comment by Swanand01** [comment]
@github-actions pre-commit

**Comment by Swanand01** [comment]
@github-actions pre-commit

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
```
 pandas/plotting/_matplotlib/timeseries.py:212:35: F821 undefined name 'DataFrame'
pandas/io/formats/info.py:112:23: F821 undefined name 'Series'
pandas/io/formats/info.py:415:35: F821 undefined name 'Series'
pandas/core/window/expanding.py:590:16: F821 undefined name 'DataFrame'
pandas/core/window/expanding.py:590:28: F821 undefined name 'Series'
pandas/core/window/expanding.py:655:16: F821 undefined name 'DataFrame'
pandas/core/window/expanding.py:655:28: F821 undefined name 'Series'
pandas/core/strings/accessor.py:2314:10: F821 undefined name 'DataFrame'
pandas/core/strings/accessor.py:2314:22: F821 undefined name 'Series'
pandas/plotting/_matplotlib/tools.py:56:15: F821 undefined name 'DataFrame'
pandas/plotting/_matplotlib/tools.py:56:27: F821 undefined name 'Series'
pandas/core/window/ewm.py:514:16: F821 undefined name 'DataFrame'
pandas/core/window/ewm.py:514:28: F821 undefined name 'Series'
pandas/core/window/ewm.py:581:16: F821 undefined name 'DataFrame'
pandas/core/window/ewm.py:581:28: F821 undefined name 'Series'
pandas/core/util/hashing.py:77:18: F821 undefined name 'DataFrame'
```

For such files you should import `Series` and `DataFrame`. If you need to avoid a circular import, you can do like they do here:

https://github.com/pandas-dev/pandas/blob/8f70ed53321642486f98c2a46d9ddfb3c80cf893/pandas/compat/pickle_compat.py#L26-L30

**Comment by Swanand01** [comment]
Hi @MarcoGorelli, would you mind taking a look at my last commit, I don't really know why the particular test is failing. 

Thank you.

**Comment by simonjayhawkins** [comment]
> Hi @MarcoGorelli, would you mind taking a look at my last commit, I don't really know why the particular test is failing.
> 
> Thank you.

see https://github.com/pandas-dev/pandas/pull/41955#pullrequestreview-682308945 for the mypy failures.

There is also a pre-commit failure. see https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit

**Comment by Swanand01** [comment]
I fixed all the mypy errors from the prev. commit in this commit. Should I run @github-actions pre-commit like it says in the documentation?

**Comment by MarcoGorelli** [comment]
> I fixed all the mypy errors from the prev. commit in this commit. 

They're still failing, if you click "details" you'll see the failures:

Screenshot 2021-06-15 at 16 10 28

> Should I run @github-actions pre-commit like it says in the documentation?

This won't fix the `flake8` errors, you'll probably need to run it locally before committing

**Comment by Swanand01** [comment]
Hi, I can't seem to understand these pre-commit errors. Are they formatting related?

**Comment by MarcoGorelli** [comment]
Hey @Swanand01 - yeah, see [here](https://pandas.pydata.org/pandas-docs/stable/development/contributing.html#pre-commit) for how to run them

Seems like this is on the right track anyway!

**Comment by Swanand01** [comment]
Hi @MarcoGorelli , all checks pass now. Thank you.

**Review by MarcoGorelli — APPROVED** [review]
Looks good to me, thanks for having stuck with this!

**Comment by Swanand01** [comment]
Hi, how will this PR be merged? Do I have to close this PR now?

**Comment by simonjayhawkins** [comment]
> Hi, how will this PR be merged? Do I have to close this PR now?

no do not close. other reviewers could either request changes, add their approval or merge this.

**Comment by Swanand01** [comment]
> > Hi, how will this PR be merged? Do I have to close this PR now?
> 
> no do not close. other reviewers could either request changes, add their approval or merge this.

Okay thanks.

File: pandas/core/frame.py [hunk 1]
```diff
@@ -1362,7 +1361,7 @@ def dot(self, other: Series) -> Series:
     def dot(self, other: DataFrame | Index | ArrayLike) -> DataFrame:
         ...
 
-    def dot(self, other: AnyArrayLike | FrameOrSeriesUnion) -> FrameOrSeriesUnion:
+    def dot(self, other: AnyArrayLike | DataFrame | Series) -> DataFrame | Series:
```
Comment (simonjayhawkins) [inline pandas/core/frame.py:1364]: A Series is also AnyArrayLike
```suggestion
    def dot(self, other: AnyArrayLike | DataFrame) -> DataFrame | Series:
```

File: pandas/core/frame.py [hunk 2]
```diff
@@ -1478,13 +1477,13 @@ def __matmul__(self, other: Series) -> Series:
 
     @overload
     def __matmul__(
-        self, other: AnyArrayLike | FrameOrSeriesUnion
-    ) -> FrameOrSeriesUnion:
+        self, other: AnyArrayLike | DataFrame | Series
```
Comment (simonjayhawkins) [inline pandas/core/frame.py:1480]: same (and already included in previous overload anyway)

File: pandas/core/frame.py [hunk 3]
```diff
@@ -1478,13 +1477,13 @@ def __matmul__(self, other: Series) -> Series:
...
+    ) -> DataFrame | Series:
         ...
 
     def __matmul__(
-        self, other: AnyArrayLike | FrameOrSeriesUnion
-    ) -> FrameOrSeriesUnion:
+        self, other: AnyArrayLike | DataFrame | Series
```
Comment (simonjayhawkins) [inline pandas/core/frame.py:1485]: same

**Review by simonjayhawkins — COMMENTED** [review]
Thanks @Swanand01 lgtm. just a couple of suggestions.

File: pandas/core/frame.py (hunk 1 above)
Comment (MarcoGorelli) [reply pandas/core/frame.py:1364]: ooh, we should add a check for this

Comment (simonjayhawkins) [reply pandas/core/frame.py:1364]: ok. we can merge this and let the check fix this

**Review by simonjayhawkins — APPROVED** [review]

**Comment by simonjayhawkins** [comment]
Thanks @Swanand01

**Comment by Dr-Irv** [comment]
@simonjayhawkins I thought we discussed in our monthly dev meeting to _not_ do this change because you lose the typeahead feature in editors.  I.e., you can do typeahead for `FrameOrSeriesUnion` but not `DataFrame | Series`

**Comment by MarcoGorelli** [comment]
@Dr-Irv my understanding was that that conversation in the dev meeting was spurred from the fact that it looked like this PR had gone stale and it was marked for 1.3. In light of that, and of the typeahead issue you brought up, it didn't seem worth pushing it though

Once activity resumed, though, it seemed fine to stick with it. Perhaps I should've pinged you to check, sorry about that

No objection to reverting (or re-discussing in the next dev meeting) if you wish, but I do think this is fine to keep. An unexpected benefit is that it's helped make it clear that there are cases when it's redundant to have `| Series`, e.g. when there's `AnyArrayLike`.
Plus, I still think it's more explicit than `FrameOrSeriesUnion` (which can be confused with `FrameOrSeries`) and that this outweights having a shorter typeahead

**Comment by simonjayhawkins** [comment]
> but I do think this is fine to keep. An unexpected benefit is that it's helped make it clear that there are cases when it's redundant to have `| Series`, e.g. when there's `AnyArrayLike`.
> Plus, I still think it's more explicit than `FrameOrSeriesUnion` (which can be confused with `FrameOrSeries`) and that this outweights having a shorter typeahead

agreed. and with #42367 removing the redundant typevars, things become clearer still.

**Pull request merged**